// ngx-ace-wrapper/fesm5/ngx-ace-wrapper.js
import { __decorate, __param } from 'tslib';
import { edit } from 'brace';
import 'brace/mode/text';
import 'brace/theme/github';
import { InjectionToken, EventEmitter, NgZone, ElementRef, KeyValueDiffers, Optional, Inject, Input, Output, Directive, HostBinding, ViewChild, Component, ViewEncapsulation, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

var ACE_CONFIG = new InjectionToken('ACE_CONFIG');
var AceEditorEvents = [
    'blur',
    'focus',
    'copy',
    'paste',
    'change',
    'changeSession'
];
var AceSelectionEvents = [
    'changeCursor',
    'changeSelection'
];
var AceConfig = /** @class */ (function () {
    function AceConfig(config) {
        if (config === void 0) { config = {}; }
        this.assign(config);
    }
    AceConfig.prototype.assign = function (config, target) {
        if (config === void 0) { config = {}; }
        target = target || this;
        for (var key in config) {
            if (config[key] && !Array.isArray(config[key]) && typeof config[key] === 'object') {
                target[key] = {};
                this.assign(config[key], target[key]);
            }
            else {
                target[key] = config[key];
            }
        }
    };
    return AceConfig;
}());

var AceDirective = /** @class */ (function () {
    function AceDirective(zone, elementRef, differs, defaults) {
        this.zone = zone;
        this.elementRef = elementRef;
        this.differs = differs;
        this.defaults = defaults;
        this.instance = null;
        this.configDiff = null;
        this.disabled = false;
        this.blur = new EventEmitter();
        this.focus = new EventEmitter();
        this.copy = new EventEmitter();
        this.paste = new EventEmitter();
        this.change = new EventEmitter();
        this.changeCursor = new EventEmitter();
        this.changeSession = new EventEmitter();
        this.changeSelection = new EventEmitter();
    }
    AceDirective.prototype.ngOnInit = function () {
        var _this = this;
        var params = new AceConfig(this.defaults);
        params.assign(this.config); // Custom configuration
        if (this.disabled) {
            params.readOnly = true;
            params.highlightActiveLine = false;
        }
        params.mode = 'ace/mode/' + (params.mode || 'text');
        params.theme = 'ace/theme/' + (params.theme || 'github');
        this.zone.runOutsideAngular(function () {
            _this.instance = edit(_this.elementRef.nativeElement);
            _this.instance.$blockScrolling = Infinity;
            _this.instance.setOptions(params);
        });
        // Add native Ace event handling
        AceEditorEvents.forEach(function (eventName) {
            if (_this.instance) {
                _this.instance.on(eventName, function () {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i] = arguments[_i];
                    }
                    if (args.length === 1) {
                        args = args[0];
                    }
                    if (_this[eventName]) {
                        _this.zone.run(function () {
                            if (_this[eventName].observers.length) {
                                _this[eventName].emit(args);
                            }
                        });
                    }
                });
            }
        });
        // Add native Ace selection event handling
        AceSelectionEvents.forEach(function (eventName) {
            if (_this.instance) {
                _this.instance.selection.on(eventName, function () {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i] = arguments[_i];
                    }
                    if (args.length === 1) {
                        args = args[0];
                    }
                    if (_this[eventName]) {
                        if (_this[eventName].observers.length) {
                            _this[eventName].emit(args);
                        }
                    }
                });
            }
        });
        if (!this.configDiff) {
            this.configDiff = this.differs.find(this.config || {}).create();
            this.configDiff.diff(this.config || {});
        }
    };
    AceDirective.prototype.ngDoCheck = function () {
        if (this.configDiff) {
            var changes = this.configDiff.diff(this.config || {});
            if (changes) {
                this.ngOnDestroy();
                this.ngOnInit();
            }
        }
    };
    AceDirective.prototype.ngOnDestroy = function () {
        if (this.instance) {
            if (this.instance.isFocused()) {
                this.blur.emit();
            }
            delete this.instance;
            this.instance = null;
        }
    };
    AceDirective.prototype.ngOnChanges = function (changes) {
        var _this = this;
        if (changes['disabled']) {
            if (changes['disabled'].currentValue !== changes['disabled'].previousValue) {
                this.zone.runOutsideAngular(function () {
                    if (_this.instance) {
                        var params = new AceConfig(_this.defaults);
                        params.assign(_this.config); // Custom configuration
                        _this.instance.clearSelection();
                        var hlActive = (params.highlightActiveLine == null) ? true : false;
                        _this.instance.setHighlightActiveLine(_this.disabled ? false : hlActive);
                        _this.instance.setReadOnly(_this.disabled ? true : (params.readOnly || false));
                    }
                });
            }
        }
    };
    AceDirective.prototype.ace = function () {
        return this.instance;
    };
    AceDirective.prototype.clear = function () {
        if (this.instance) {
            this.instance.setValue('');
            this.instance.clearSelection();
        }
    };
    AceDirective.prototype.getValue = function () {
        if (this.instance) {
            return this.instance.getValue();
        }
    };
    AceDirective.prototype.setValue = function (value, cursorPos) {
        if (this.instance) {
            this.instance.setValue(value || '', cursorPos);
        }
    };
    AceDirective.ctorParameters = function () { return [
        { type: NgZone },
        { type: ElementRef },
        { type: KeyValueDiffers },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [ACE_CONFIG,] }] }
    ]; };
    __decorate([
        Input()
    ], AceDirective.prototype, "disabled", void 0);
    __decorate([
        Input('ace')
    ], AceDirective.prototype, "config", void 0);
    __decorate([
        Output()
    ], AceDirective.prototype, "blur", void 0);
    __decorate([
        Output()
    ], AceDirective.prototype, "focus", void 0);
    __decorate([
        Output()
    ], AceDirective.prototype, "copy", void 0);
    __decorate([
        Output()
    ], AceDirective.prototype, "paste", void 0);
    __decorate([
        Output()
    ], AceDirective.prototype, "change", void 0);
    __decorate([
        Output()
    ], AceDirective.prototype, "changeCursor", void 0);
    __decorate([
        Output()
    ], AceDirective.prototype, "changeSession", void 0);
    __decorate([
        Output()
    ], AceDirective.prototype, "changeSelection", void 0);
    AceDirective = __decorate([
        Directive({
            selector: '[ace]',
            exportAs: 'ngxAce'
        }),
        __param(3, Optional()), __param(3, Inject(ACE_CONFIG))
    ], AceDirective);
    return AceDirective;
}());

var AceComponent = /** @class */ (function () {
    function AceComponent() {
        this.content = '';
        this.disabled = false;
        this.mode = '';
        this.theme = '';
        this.useAceClass = true;
        this.blur = new EventEmitter();
        this.focus = new EventEmitter();
        this.copy = new EventEmitter();
        this.paste = new EventEmitter();
        this.change = new EventEmitter();
        this.valueChange = new EventEmitter();
        this.changeCursor = new EventEmitter();
        this.changeSession = new EventEmitter();
        this.changeSelection = new EventEmitter();
    }
    Object.defineProperty(AceComponent.prototype, "value", {
        get: function () {
            return this.content;
        },
        set: function (value) {
            this.setContent(value);
        },
        enumerable: true,
        configurable: true
    });
    AceComponent.prototype.ngAfterViewInit = function () {
        if (this.content != null) {
            this.setContent(this.content, true);
        }
    };
    AceComponent.prototype.setContent = function (value, force) {
        if (force || value !== this.content) {
            if (this.directiveRef) {
                this.directiveRef.setValue(value, 1);
            }
            this.content = value;
        }
    };
    AceComponent.prototype.getConfig = function () {
        this.config = this.config || {};
        this.config.mode = this.mode || this.config.mode;
        this.config.theme = this.theme || this.config.theme;
        return this.config;
    };
    AceComponent.prototype.onContentChange = function (event) {
        if (this.directiveRef) {
            this.change.emit(event);
            this.content = this.directiveRef.getValue() || '';
            this.valueChange.emit(this.value);
        }
    };
    __decorate([
        Input()
    ], AceComponent.prototype, "value", null);
    __decorate([
        Input()
    ], AceComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], AceComponent.prototype, "mode", void 0);
    __decorate([
        Input()
    ], AceComponent.prototype, "theme", void 0);
    __decorate([
        Input()
    ], AceComponent.prototype, "config", void 0);
    __decorate([
        HostBinding('class.ace'),
        Input()
    ], AceComponent.prototype, "useAceClass", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "blur", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "focus", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "copy", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "paste", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "change", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "valueChange", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "changeCursor", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "changeSession", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "changeSelection", void 0);
    __decorate([
        ViewChild(AceDirective, { static: true })
    ], AceComponent.prototype, "directiveRef", void 0);
    AceComponent = __decorate([
        Component({
            selector: 'ace',
            exportAs: 'ngxAce',
            template: "<div [ace]=\"getConfig()\" [disabled]=\"disabled\" (blur)=\"blur.emit($event)\" (focus)=\"focus.emit($event)\" (copy)=\"copy.emit($event)\" (paste)=\"paste.emit($event)\" (change)=\"onContentChange($event)\" (changeCursor)=\"changeCursor.emit($event)\" (changeSession)=\"changeSession.emit($event)\" (changeSelection)=\"changeSelection.emit($event)\">\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            styles: [".ace,.ace>.ace_editor{display:block;width:100%;height:100%}.ace[fxflex]{display:-webkit-box;display:flex;flex-direction:inherit;-webkit-box-align:inherit;align-items:inherit;min-width:0;min-height:0;-webkit-box-direction:inherit;-webkit-box-orient:inherit}.ace[fxflex]>.ace_editor{flex:1 1 auto;min-width:0;min-height:0;-webkit-box-flex:1}"]
        })
    ], AceComponent);
    return AceComponent;
}());

var AceModule = /** @class */ (function () {
    function AceModule() {
    }
    AceModule = __decorate([
        NgModule({
            declarations: [AceComponent, AceDirective],
            imports: [CommonModule],
            exports: [CommonModule, AceComponent, AceDirective],
            providers: []
        })
    ], AceModule);
    return AceModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { ACE_CONFIG, AceComponent, AceConfig, AceDirective, AceModule };
