// ngx-ace-wrapper/esm5/lib/ace.component.js
import { __decorate } from "tslib";
import 'brace';
import 'brace/mode/text';
import 'brace/theme/github';
import { Component, Input, Output, EventEmitter, ViewChild, HostBinding, ViewEncapsulation } from '@angular/core';
import { AceDirective } from './ace.directive';
var AceComponent = /** @class */ (function () {
    function AceComponent() {
        this.content = '';
        this.disabled = false;
        this.mode = '';
        this.theme = '';
        this.useAceClass = true;
        this.blur = new EventEmitter();
        this.focus = new EventEmitter();
        this.copy = new EventEmitter();
        this.paste = new EventEmitter();
        this.change = new EventEmitter();
        this.valueChange = new EventEmitter();
        this.changeCursor = new EventEmitter();
        this.changeSession = new EventEmitter();
        this.changeSelection = new EventEmitter();
    }
    Object.defineProperty(AceComponent.prototype, "value", {
        get: function () {
            return this.content;
        },
        set: function (value) {
            this.setContent(value);
        },
        enumerable: true,
        configurable: true
    });
    AceComponent.prototype.ngAfterViewInit = function () {
        if (this.content != null) {
            this.setContent(this.content, true);
        }
    };
    AceComponent.prototype.setContent = function (value, force) {
        if (force || value !== this.content) {
            if (this.directiveRef) {
                this.directiveRef.setValue(value, 1);
            }
            this.content = value;
        }
    };
    AceComponent.prototype.getConfig = function () {
        this.config = this.config || {};
        this.config.mode = this.mode || this.config.mode;
        this.config.theme = this.theme || this.config.theme;
        return this.config;
    };
    AceComponent.prototype.onContentChange = function (event) {
        if (this.directiveRef) {
            this.change.emit(event);
            this.content = this.directiveRef.getValue() || '';
            this.valueChange.emit(this.value);
        }
    };
    __decorate([
        Input()
    ], AceComponent.prototype, "value", null);
    __decorate([
        Input()
    ], AceComponent.prototype, "disabled", void 0);
    __decorate([
        Input()
    ], AceComponent.prototype, "mode", void 0);
    __decorate([
        Input()
    ], AceComponent.prototype, "theme", void 0);
    __decorate([
        Input()
    ], AceComponent.prototype, "config", void 0);
    __decorate([
        HostBinding('class.ace'),
        Input()
    ], AceComponent.prototype, "useAceClass", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "blur", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "focus", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "copy", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "paste", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "change", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "valueChange", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "changeCursor", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "changeSession", void 0);
    __decorate([
        Output()
    ], AceComponent.prototype, "changeSelection", void 0);
    __decorate([
        ViewChild(AceDirective, { static: true })
    ], AceComponent.prototype, "directiveRef", void 0);
    AceComponent = __decorate([
        Component({
            selector: 'ace',
            exportAs: 'ngxAce',
            template: "<div [ace]=\"getConfig()\" [disabled]=\"disabled\" (blur)=\"blur.emit($event)\" (focus)=\"focus.emit($event)\" (copy)=\"copy.emit($event)\" (paste)=\"paste.emit($event)\" (change)=\"onContentChange($event)\" (changeCursor)=\"changeCursor.emit($event)\" (changeSession)=\"changeSession.emit($event)\" (changeSelection)=\"changeSelection.emit($event)\">\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            styles: [".ace,.ace>.ace_editor{display:block;width:100%;height:100%}.ace[fxflex]{display:-webkit-box;display:flex;flex-direction:inherit;-webkit-box-align:inherit;align-items:inherit;min-width:0;min-height:0;-webkit-box-direction:inherit;-webkit-box-orient:inherit}.ace[fxflex]>.ace_editor{flex:1 1 auto;min-width:0;min-height:0;-webkit-box-flex:1}"]
        })
    ], AceComponent);
    return AceComponent;
}());
export { AceComponent };
